azureeyemodule/app/iot/iot_interface.cpp

```diff
@@ -147,62 +149,104 @@ static inline std::unordered_map<std::string, std::string> parse_onvif_message(c
+                {
+                    case(0):
+                        curtok = part;
+                        break;
+                    case(1):
+                        result[curtok] = part;
+                        break;
+                    case(2):
+                    // the only feature by far take in two parameters is snapshot
+                        if (curtok != "snapshot") 
+                        {
+                            util::log_error("Malformed ONVIF control message: " +msg);
+                        }
+                        //result[curtok] = result[curtok] + "/" + part;
+                        result[curtok] = part;
+                        break;
+                    default:
+                        util::log_error("Malformed ONVIF control message: " +msg);
+                }
+            }
+            tokens_and_values[i].erase(0, pos + 1);
+            count += 1;
         }
     }
-
-    // If there are not an even number of items in the vector, the message was malformed.
-    if ((tokens_and_values.size() % 2) != 0)
-    {
-        util::log_error("Malformed ONVIF control message. Will try to work with it: " + msg);
-        result[curtok] = std::string("");
-    }
-
     return result;
 }
 
 /** RECV_MSG message handler. Handles incoming messages for the ONVIF channel. */
 static IOTHUBMESSAGE_DISPOSITION_RESULT receive_onvif_message(IOTHUB_MESSAGE_HANDLE message, void *unused)
 {
-    const std::string msg = std::string(IoTHubMessage_GetMessageId(message));
-    util::log_info("Got an ONVIF control message: " + msg);
-
-    auto deserialized_msg = parse_onvif_message(msg);
+    const unsigned char* buffer = 0;
+    size_t s = 256;
+    IoTHubMessage_GetByteArray(message, &buffer, &s);
+    char str[s +1];
+    memcpy(str, buffer, s);
+    str[s] = 0;
+    JSON_Value *root_value = json_parse_string(reinterpret_cast<const char*>(str));
+    JSON_Object *root_object = json_value_get_object(root_value);
+    util::log_info("Got an ONVIF control message: " + std::string(str));
 
     // Set stuff based on results
-    if (deserialized_msg.find(std::string("fps")) != deserialized_msg.end())
+    if (json_object_get_value(root_object, "snapshot") != nullptr)
+    {
+        printf("accept taking snapshot \n\n");
+        std::string type  = json_object_get_string(root_object, "snapshot");
+        if (type == "raw")
+        {
+            rtsp::take_snapshot(rtsp::StreamType::RAW);
+        }
+        else if ("result")
+        {
+            rtsp::take_snapshot(rtsp::StreamType::RESULT);
+        }
+        else
+        {
+            util::log_error("receive a invalid snapshot type request.");
+        }
+    }
+           
+
+    if (json_object_get_value(root_object, "-fps") != nullptr)
     {
-        const auto value = deserialized_msg[std::string("fps")];
+        printf("accept frame per second change \n\n");
+        std::string value  = json_object_get_string(root_object, "-fps");
         try
```

Comment:
Why are we "try"ing this? Should there be a corresponding "catch"?

Reply:
Hi, this "try" is needed. I need a way to figure out if the fps parameter is a proper integer. The following "catch" would catch the error if the parameter is unable to be converted into integer

Reply:
That's fine, but I don't actually see a "catch" anywhere. Try not to catch all exceptions - just the one or two that you are actually looking for, then log an error that explains that the integer is not valid. Looks like: http://www.cplusplus.com/reference/string/stoi/, you are probably just looking for an invalid_argument exception.

Reply:
The catch line is over here (line 9), this is the complete code:

```
if (json_object_get_value(root_object, "-fps") != nullptr)
    {
        std::string value  = json_object_get_string(root_object, "-fps");
        try
        {
            rtsp::set_stream_params(rtsp::StreamType::RAW, std::stoi(value));
            rtsp::set_stream_params(rtsp::StreamType::RESULT, std::stoi(value));
        }
        catch (std::invalid_argument &e)
        {
            util::log_error("While parsing ONVIF message, could not convert FPS parameter into an integer: " + value);
            return IOTHUBMESSAGE_REJECTED;
        }
    }
```

Reply:
Yup, you are right. GitHub was being coy with me 😠